### PR TITLE
feat!(train): Loss/eval基準でもモデルを残すように

### PR DESCRIPTION
**Pull request by sevenc-nanashi**
## 内容

モデルを残すコードを共通化し、Loss/eval基準でもモデルを残すようにします。

## 関連 Issue

- closes: #106

## スクリーンショット・動画など

（なし）

## その他

共通化の都合上、最新epochのモデルの名前が変わるので、 #112 が先。

### CODE REVIEW

File: train/src/train.py [hunk 1]
```diff
@@ -31,6 +32,54 @@ def word_to_tensor(word: str, device: torch.device) -> torch.Tensor:
...
+                print(f"[{self.label}] Will not save the current model")
+                return
+            else:
+                path = self.output_dir / f"model-{self.label}-e{removed_epoch}.pth"
+                print(f"[{self.label}] Removing {path}")
+                os.remove(path)
+
+        path = self.output_dir / f"model-{self.label}-e{current_epoch}.pth"
+        print(f"[{self.label}] Saving {path}")
+        torch.save(self.model.state_dict(), path)
```
Comment (Hiroshiba) [inline train/src/train.py:80]: （ただのコメントです）

listやファイルIOなど至るところにステートがあって、今後この関数に機能追加するとバグりやすそうなメンテナンス性低めのコードかもなーと感じました。
例えば低いスコアのmodelでも一度`models`に足してからpopしてるとか。

といっても良いリファクタリングを思いつかなかったので、ただのコメントだけです。。。
今後機能追加する前にリファクタリングしたほうが良さそうということで、ここに`#TODO: リファクタリングする`みたいなの書いて後の自分がわかるようにしておくとかでも良いのかも。

**Review by Hiroshiba — APPROVED** [review]
LGTM！！

いくつかコメントしていますが、まあ別に強い意見ではないかなーという気持ちです！

File: train/src/train.py [hunk 2]
```diff
@@ -31,6 +32,54 @@ def word_to_tensor(word: str, device: torch.device) -> torch.Tensor:
     return torch.tensor(indices, device=device)
 
 
+class ModelKeeper:
```
Comment (Hiroshiba) [inline train/src/train.py:35]: （強くないコメントです）

`ModelKeeper`と`step`だと何をするのか少しピンと来にくいかもと思いました。
といっても最高のアイデアはないのですが。。

よく`checkpoint`と呼んだりするので、`ModelCheckpoint`や`CheckpointManager`にすると機械学習勢には伝わりやすいかもです。
あるいは｀BestModelManager`とか`BestModelSaver`ですかね。。。
（保存するのはBestだけではありませんが。。）

`step`だとなにするか分かりづらいけど、かといって最高な関数名はなさそう。。。
`update`だと状態を更新するのがすこしわかりやすい･･････かも、くらい。

Comment (sevenc-nanashi) [reply train/src/train.py]: updateにしました。

**Comment by Hiroshiba** [comment]
ちょっとリファクタリングコード書いてみました、参考になれば！！

```py
class ModelKeeper:
    def __init__(
        self,
        label: str,
        model: nn.Module,
        output_dir: Path,
        num_models_to_keep: int,
        compare_mode: Literal["max", "min"],
    ) -> None:
        self.label = label
        self.model = model
        self.output_dir = output_dir
        self.output_dir.mkdir(parents=True, exist_ok=True)

        self._limit = num_models_to_keep
        self._compare_mode = compare_mode
        self._models: dict[int, float] = {}

    def step(self, epoch: int, score: int | float) -> None:
        if len(self._models) < self._limit:
            return

        worst_epoch = self._worst({**self._models, epoch: score})
        if worst_epoch == epoch:
            return

        self._models.pop(worst_epoch)
        self._models[epoch] = score

        self._path(worst_epoch).unlink()
        torch.save(self.model.state_dict(), self._path(epoch))

    def _worst(self, models: dict[int, float]) -> int:
        _factor = 1 if self._compare_mode == "max" else -1
        return min(models.items(), key=lambda x: x[1] * _factor)[0]

    def _path(self, epoch: int) -> Path:
        return self.output_dir / f"model-{self.label}-e{epoch}.pth"
```

**Comment by Hiroshiba** [comment]
んまあTODOコメント書いてマージでも良いと思います！

**Comment by sevenc-nanashi** [comment]
置き換えたのでマージします。
